rascal-lsp/src/main/rascal/lang/rascal/lsp/Actions.rsc

```diff
@@ -0,0 +1,156 @@
+@license{
+Copyright (c) 2018-2023, NWO-I CWI and Swat.engineering
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met:
+
+1. Redistributions of source code must retain the above copyright notice,
+this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright notice,
+this list of conditions and the following disclaimer in the documentation
+and/or other materials provided with the distribution.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+}
+@bootstrapParser
+@synopsis{Defines both the command evaluator and the codeAction retriever for Rascal}
+module lang::rascal::lsp::Actions
+
+import lang::rascal::\syntax::Rascal;
+import util::LanguageServer;
+import analysis::diff::edits::TextEdits;
+import ParseTree;
+import String;
+import lang::rascal::vis::ImportGraph;
+import util::Reflective;
+import util::IDEServices;
+import List;
+import IO;
+
+@synopsis{Here we list Rascal-specific code commands}
+@description{
+The commands must be evaluated by ((evaluateRascalCommand))
+}
+data Command
+    = visualImportGraphCommand(PathConfig pcfg)
+    | sortImportsAndExtends(Header h)
+    ;
+
+@synopsis{Detects (on-demand) source actions to register with specific places near the current cursor}
+list[CodeAction] rascalCodeActions(Focus focus, PathConfig pcfg=pathConfig()) {
+    result = [];
+
+    if ([*_, start[Module] top] := focus) {
+        result += addLicenseAction(top, pcfg);
+    }
+
+    if ([*_, Toplevel t, *_] := focus) {
+        result += toplevelCodeActions(t);
+    }
+
+    if ([*_, Header h, *_] := focus) {
+        result += [action(command=visualImportGraphCommand(pcfg), title="Visualize project import graph")]
+               +  [action(command=sortImportsAndExtends(h), title="Sort imports and extends")]
+               ;
+    }
+
+    return result;
+}
+
+@synopsis{Add a license header if there isn't one.}
+list[CodeAction] addLicenseAction(start[Module] \module, PathConfig pcfg) {
```

Comment:
not everyone wants to put a license header in their source code. Some licenses allow for only a license in the root of the directory. Can a user disable these in some way?

Reply:
You can simply not choose to activate quickfix? If my memory does not fail me, the fact that licenses can only be in the root of the folder is less depending on the license and more on the locality and the scope of the license. American law is quite different from European law, and anglosaxon law from continental European law. 

In any case, the prudent lawyer would advice you to put it everywhere if you (a) don't want to be liable for the contents if someone copies an entire file without the project contents and (b) want the copier of the file to be liable to the conditions made in the license. 

In that sense this action is useful to anybody who wishes to be on the safe side.

Reply:
I've restricted the action to only appear if the cursor is on the QualifiedName part of the header (the module name). That way it is always within reach but not annoyingly so.